### ISSUE TITLE
Some confusions about the config module

### ISSUE THREAD

**Issue by suchow** [Dallinger/Dallinger#434]
# What I would have expected

Both in the context of a standalone script and a running experiment, the following can be used to access and set configuration variables:

```
from dallinger import config

config.get("foo")
config.set("foo", "bar")
```

When imported in the context of an experiment, the set of valid keys is extended to include those defined by the experiment. In all cases, the existing precedence ranking for configuration is in place (environment > config.txt > .dallingerconfig > defaults).

# What I observe

I was trying to use access a configuration variable when writing a test for the Open Science Framework and ran into the following issues:

## 1

To use `get_config`, it must be imported in a particular way. 

This is okay:

```
from dallinger.config import get_config

config = get_config()
```

But this is not:

```
import dallinger

config = dallinger.config.get_config()
```

Nor is this:

```
from dallinger import config

config = config.get_config()
```
In both cases it leads to the following error:
```
❯ python tests/test_registration.py
Traceback (most recent call last):
  File "tests/test_registration.py", line 4, in <module>
    config = dallinger.config.get_config()
TypeError: 'NoneType' object is not callable
```

## 2

The config object is returned before it is loaded and ready to read from:

```

from dallinger.config import get_config

config = get_config()
config.get("osf_access_token")
```

```
❯ python tests/test_registration.py
Traceback (most recent call last):
  File "tests/test_registration.py", line 5, in <module>
    config.get("osf_access_token")
  File "/Users/jordansuchow/Dropbox/Berkeley/Projects/Current/Dallinger/dallinger/config.py", line 67, in get
    raise RuntimeError('Config loading not finished')
RuntimeError: Config loading not finished
```

## 3 

Loading the config explicitly fixes 2, which is confusing because I would have thought then that the error message above should have been `Config not loaded`.
```
from dallinger.config import get_config

config = get_config()
config.load_config()

config.get("osf_access_token")
```

## 4

In the context of a debugged experiment, this isn't an issue and I'm able to access config variables without loading the config object. (Good.)

**Comment by alecpm** [comment]
I've pushed fixes for the above issues to the development branch. The import issue was caused by an old instance of a `LocalConfig` config object that was instantiated at the module level in dallinger. I believe that `config` object had been retained for some backwards compatibility that is no longer necessary. The new `config` module should now be importable in the manner you suggest above.

I've changed the error message to `Config not loaded`, since as you say that is a more accurate description of the condition.

In general, I believe the intent is to have something like `config = get_config()` called at module scope to make the global configuration object available to functions and methods within the module. 

Generally, it shouldn't be necessary to call `load_config` explicitly, and should probably be considered an anti-pattern, because `load_config` can generally only be called once (until/unless the `ready` flag is reset), and it needs to be called within a specific working directory. Generally, the command line tools, the experiment server, and the python experiment API should all ensure that `load_config` is called at the appropriate times and locations. Custom runtime configuration should be passed to the experiment's `sandbox` or `deploy` methods, rather than managed via explicit calls to configuration methods. Is that about right @MatthewWilkes?
